File: .archive/cli-ts/commands/branch.ts
/**
 * `agt branch` — Create a git worktree + branch for a todo.
 *
 * Creates .worktrees/PREFIX-N-title-slug/ with a new branch.
 * Sets the todo status to in_progress and stores the branch name.
 */

import { join } from "node:path";
import { existsSync } from "node:fs";
import type { Command } from "commander";
import { findProject, readConfig } from "../../lib/project.js";
import { loadDoc, saveDoc } from "../../lib/storage.js";
import { setBranch, clearBranch } from "../../lib/operations.js";
import { parseTodoRef, findTodoByNumber } from "../../lib/queries.js";
import { error, success, warn } from "../output.js";

const ARTICLES = new Set(["a", "an", "the"]);

function slugify(text: string, maxWords = 5): string {
  return text
    .toLowerCase()
    .split(/[^a-z0-9]+/)
    .filter((w) => w && !ARTICLES.has(w))
    .slice(0, maxWords)
    .join("-");
}

export function registerBranch(program: Command): void {
  program
    .command("branch")
    .description("Create a git worktree + branch for a todo")
    .argument("<ref>", "Todo reference (e.g. ABC-1 or 1)")
    .option("--no-worktree", "Create branch only, skip worktree")
    .option("--json", "Output as JSON")
    .action(async (ref: string, opts: { worktree: boolean; json?: boolean }) => {
      const paths = findProject();
      if (!paths) error("Not in an agt project. Run 'agt init' first.");

      const config = await readConfig(paths.configPath);
      let doc = await loadDoc(paths.dataPath);
      if (!doc) error("Project data not found.");

      const num = parseTodoRef(ref, config.prefix);
      if (num === null) error(`Invalid todo reference: "${ref}".`);

      const todo = findTodoByNumber(doc, num);
      if (!todo) error(`Todo ${config.prefix}-${num} not found.`);

      // Check if already branched
      if (todo.branch) {
        if (opts.json) {
          console.log(
            JSON.stringify({
              ref: `${config.prefix}-${num}`,
              branch: todo.branch,
              alreadyExists: true,
            }),
          );
        } else {
          warn(`Todo ${config.prefix}-${num} already has branch: ${todo.branch}`);
        }
        return;
      }

      const slug = slugify(todo.title);
      const branchName = `${config.prefix.toLowerCase()}-${num}-${slug}`;

      if (opts.worktree) {
        const worktreesDir = join(paths.root, ".worktrees");
        const worktreePath = join(worktreesDir, branchName);

        if (existsSync(worktreePath)) {
          error(`Worktree path already exists: ${worktreePath}`);
        }

        // Ensure .worktrees/ is in .gitignore
        const gitignorePath = join(paths.root, ".gitignore");
        const gitignoreFile = Bun.file(gitignorePath);
        if (await gitignoreFile.exists()) {
          const content = await gitignoreFile.text();
          if (!content.includes(".worktrees")) {
            await Bun.write(
              gitignorePath,
              content.trimEnd() + "\n\n# Git worktrees for todo branches\n.worktrees/\n",
            );
          }
        } else {
          await Bun.write(gitignorePath, "# Git worktrees for todo branches\n.worktrees/\n");
        }

        const result = Bun.spawnSync(["git", "worktree", "add", "-b", branchName, worktreePath], {
          cwd: paths.root,
          stderr: "pipe",
          stdout: "pipe",
        });

        if (result.exitCode !== 0) {
          const stderr = result.stderr.toString().trim();
          error(`Failed to create worktree: ${stderr}`);
        }
      } else {
        const result = Bun.spawnSync(["git", "branch", branchName], {
          cwd: paths.root,
          stderr: "pipe",
          stdout: "pipe",
        });

        if (result.exitCode !== 0) {
          const stderr = result.stderr.toString().trim();
          error(`Failed to create branch: ${stderr}`);
        }
      }

      // Store branch name on the todo
      doc = setBranch(doc, num, branchName);
      await saveDoc(paths.dataPath, doc);

      const todoRef = `${config.prefix}-${num}`;
      if (opts.json) {
        console.log(
          JSON.stringify({
            ref: todoRef,
            branch: branchName,
            worktree: opts.worktree ? `.worktrees/${branchName}` : null,
          }),
        );
      } else {
        success(`Created branch: ${branchName}`);
        if (opts.worktree) {
          console.log(`  Worktree: .worktrees/${branchName}/`);
        }
      }
    });

  program
    .command("unbranch")
    .description("Remove a git worktree + branch for a todo")
    .argument("<ref>", "Todo reference (e.g. ABC-1 or 1)")
    .option("--keep-branch", "Remove worktree but keep the git branch")
    .option("--json", "Output as JSON")
    .action(async (ref: string, opts: { keepBranch?: boolean; json?: boolean }) => {
      const paths = findProject();
      if (!paths) error("Not in an agt project. Run 'agt init' first.");

      const config = await readConfig(paths.configPath);
      let doc = await loadDoc(paths.dataPath);
      if (!doc) error("Project data not found.");

      const num = parseTodoRef(ref, config.prefix);
      if (num === null) error(`Invalid todo reference: "${ref}".`);

      const todo = findTodoByNumber(doc, num);
      if (!todo) error(`Todo ${config.prefix}-${num} not found.`);

      if (!todo.branch) {
        if (opts.json) {
          console.log(JSON.stringify({ ref: `${config.prefix}-${num}`, error: "no branch" }));
        } else {
          warn(`Todo ${config.prefix}-${num} has no branch.`);
        }
        return;
      }

      const branchName = todo.branch;
      const worktreePath = join(paths.root, ".worktrees", branchName);

      // Remove worktree if it exists
      if (existsSync(worktreePath)) {
        const result = Bun.spawnSync(["git", "worktree", "remove", worktreePath], {
          cwd: paths.root,
          stderr: "pipe",
          stdout: "pipe",
        });

        if (result.exitCode !== 0) {
          const stderr = result.stderr.toString().trim();
          error(`Failed to remove worktree: ${stderr}`);
        }
      }

      // Delete the git branch unless --keep-branch
      if (!opts.keepBranch) {
        const result = Bun.spawnSync(["git", "branch", "-d", branchName], {
          cwd: paths.root,
          stderr: "pipe",
          stdout: "pipe",
        });

        if (result.exitCode !== 0) {
          const stderr = result.stderr.toString().trim();
          // Not fatal — branch may have been deleted already or not yet merged
          warn(`Could not delete branch: ${stderr}`);
        }
      }

      // Clear branch reference on the todo
      doc = clearBranch(doc, num);
      await saveDoc(paths.dataPath, doc);

      const todoRef = `${config.prefix}-${num}`;
      if (opts.json) {
        console.log(JSON.stringify({ ref: todoRef, branch: branchName, removed: true }));
      } else {
        success(`Removed branch: ${branchName}`);
      }
    });
}

File: src/rust/crates/agt-cli/src/commands/branch.rs
//! `agt branch` — Create a git worktree + branch for a todo.
//!
//! Creates .worktrees/<branch-name>/ with a new branch.
//! Sets the branch name on the todo in the CRDT.

use anyhow::{Context, Result, bail};
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use agt_lib::operations;
use agt_lib::queries;
use agt_lib::schema::Todo;

use super::{load_project, parse_ref, save_project};

/// Articles to strip from branch slugs.
const ARTICLES: &[&str] = &["a", "an", "the"];

/// Generate a branch name from a todo: `prefix-N-title-slug`.
pub fn make_branch_name(prefix: &str, num: u64, title: &str) -> String {
    let words: Vec<String> = title
        .to_lowercase()
        .split(|c: char| !c.is_alphanumeric())
        .filter(|w| !w.is_empty() && !ARTICLES.contains(w))
        .take(5)
        .map(|s| s.to_string())
        .collect();

    let slug = words.join("-");
    format!("{}-{}-{}", prefix.to_lowercase(), num, slug)
}

/// Ensure .worktrees/ is listed in .gitignore.
fn ensure_gitignore_worktrees(root: &Path) -> Result<()> {
    let gitignore_path = root.join(".gitignore");

    if gitignore_path.exists() {
        let content = fs::read_to_string(&gitignore_path)?;
        if !content.contains(".worktrees") {
            let updated = format!(
                "{}\n\n# Git worktrees for todo branches\n.worktrees/\n",
                content.trim_end()
            );
            fs::write(&gitignore_path, updated)?;
        }
    } else {
        fs::write(
            &gitignore_path,
            "# Git worktrees for todo branches\n.worktrees/\n",
        )?;
    }
    Ok(())
}

/// Create a git worktree for a todo, returning the worktree path and branch name.
/// If the todo already has a branch, returns its existing worktree path.
///
/// This is the shared function used by both `agt branch` and `agt run`.
pub fn ensure_worktree(root: &Path, prefix: &str, todo: &Todo) -> Result<(String, PathBuf)> {
    // If already branched, return existing info
    if let Some(ref branch) = todo.branch {
        let worktree_path = root.join(".worktrees").join(branch);
        if worktree_path.exists() {
            return Ok((branch.clone(), worktree_path));
        }
        // Branch exists on todo but worktree dir is missing — recreate it
        ensure_gitignore_worktrees(root)?;
        let output = Command::new("git")
            .args(["worktree", "add", worktree_path.to_str().unwrap(), branch])
            .current_dir(root)
            .output()
            .context("Failed to run git worktree add")?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            bail!(
                "Failed to recreate worktree for existing branch {}: {}",
                branch,
                stderr.trim()
            );
        }
        return Ok((branch.clone(), worktree_path));
    }

    // Generate new branch name
    let branch_name = make_branch_name(prefix, todo.number, &todo.title);
    let worktrees_dir = root.join(".worktrees");
    let worktree_path = worktrees_dir.join(&branch_name);

    if worktree_path.exists() {
        // Worktree dir exists but todo doesn't have branch set — it's from a previous run
        return Ok((branch_name, worktree_path));
    }

    ensure_gitignore_worktrees(root)?;

    let output = Command::new("git")
        .args([
            "worktree",
            "add",
            "-b",
            &branch_name,
            worktree_path.to_str().unwrap(),
        ])
        .current_dir(root)
        .output()
        .context("Failed to run git worktree add")?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        bail!("Failed to create worktree: {}", stderr.trim());
    }

    Ok((branch_name, worktree_path))
}

/// `agt branch <ref>` — create worktree + branch, record on todo.
pub fn run(reference: String, json: bool) -> Result<()> {
    let (paths, mut doc) = load_project()?;
    let (_, prefix, _, _) = queries::read_project_meta(&doc);
    let num = parse_ref(&reference, &prefix)?;

    let todo = queries::find_todo_by_number(&doc, num)
        .ok_or_else(|| anyhow::anyhow!("Todo {}-{} not found", prefix, num))?;

    if let Some(ref branch) = todo.branch {
        let worktree_path = paths.root.join(".worktrees").join(branch);
        if json {
            println!(
                "{}",
                serde_json::json!({
                    "ok": true,
                    "reference": format!("{}-{}", prefix, num),
                    "branch": branch,
                    "worktree": worktree_path.display().to_string(),
                })
            );
        } else {
            println!("Todo {}-{} already has branch: {}", prefix, num, branch);
        }
        return Ok(());
    }

    let (branch_name, worktree_path) = ensure_worktree(&paths.root, &prefix, &todo)?;

    // Record branch on todo
    operations::set_branch(&mut doc, num, &branch_name, None)?;
    save_project(&paths, &mut doc)?;

    if json {
        println!(
            "{}",
            serde_json::json!({
                "ok": true,
                "reference": format!("{}-{}", prefix, num),
                "branch": branch_name,
                "worktree": worktree_path.display().to_string(),
            })
        );
    } else {
        println!("Created branch: {}", branch_name);
        println!("  Worktree: {}", worktree_path.display());
    }
    Ok(())
}
